test/hotspot/jtreg/serviceability/jvmti/ObjectMonitorUsage/libObjectMonitorUsage.cpp

```diff
@@ -0,0 +1,249 @@
+/*
+ * Copyright (c) 2024, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+#include <stdio.h>
+#include <string.h>
+#include "jvmti.h"
+#include "jvmti_common.hpp"
+
+extern "C" {
+
+#define PASSED 0
+#define STATUS_FAILED 2
+
+static jvmtiEnv *jvmti = nullptr;
+static jrawMonitorID event_lock = nullptr;
+static jint result = PASSED;
+static int check_idx = 0;
+static int waits_to_enter = 0;
+static int waits_to_be_notified = 0;
+static jobject tested_monitor = nullptr;
+
+static bool is_tested_monitor(JNIEnv *jni, jobject monitor) {
+  if (tested_monitor == nullptr) {
+    return false; // tested_monitor was not set yet
+  }
+  return jni->IsSameObject(monitor, tested_monitor) == JNI_TRUE;
+}
+
+JNIEXPORT void JNICALL
+MonitorContendedEnter(jvmtiEnv *jvmti, JNIEnv *jni, jthread thread, jobject monitor) {
+  RawMonitorLocker rml(jvmti, jni, event_lock);
+  if (is_tested_monitor(jni, monitor)) {
+    waits_to_enter++;
+  }
+}
+
+JNIEXPORT void JNICALL
+MonitorContendedEntered(jvmtiEnv *jvmti, JNIEnv *jni, jthread thread, jobject monitor) {
+  RawMonitorLocker rml(jvmti, jni, event_lock);
+  if (is_tested_monitor(jni, monitor)) {
+    waits_to_enter--;
+  }
+}
+
+JNIEXPORT void JNICALL
+MonitorWait(jvmtiEnv *jvmti, JNIEnv *jni, jthread thread, jobject monitor, jlong timeout) {
+  RawMonitorLocker rml(jvmti, jni, event_lock);
+  if (is_tested_monitor(jni, monitor)) {
+    waits_to_be_notified++;
+  }
+}
+
+JNIEXPORT void JNICALL
+MonitorWaited(jvmtiEnv *jvmti, JNIEnv *jni, jthread thread, jobject monitor, jboolean timed_out) {
+  RawMonitorLocker rml(jvmti, jni, event_lock);
+  if (is_tested_monitor(jni, monitor)) {
+    waits_to_be_notified--;
+  }
+}
```

Comment:
I like these for getting rid of the raciness of knowing when a thread is blocked
on contention, finished entering, blocked on wait or finished waiting. Nice!